libvast/src/table_slice.cpp

```diff
@@ -891,6 +883,263 @@ auto resolve_operand(const table_slice& slice, const operand& op)
+  return result;
+}
+
+auto make_flatten_transformation(
+  const flatten_options& opt, const std::string& name_prefix,
+  std::vector<std::shared_ptr<arrow::Array>> list_offsets)
+  -> indexed_transformation::function_type {
+  return [=](struct record_type::field field,
+             std::shared_ptr<arrow::Array> array)
+           -> indexed_transformation::result_type {
+    auto f = detail::overload{
+      [&]<concrete_type Type>(
+        const Type&) -> indexed_transformation::result_type {
+        // Return unchanged, but use prefix, and wrap in a list if we need to.
+        if (list_offsets.empty()) {
+          return {
+            {
+              {
+                fmt::format("{}{}", name_prefix, field.name),
+                field.type,
+              },
+              array,
+            },
+          };
+        }
+        auto combined_offsets = combine_offsets(list_offsets);
+        return {
+          {
+            {
+              fmt::format("{}{}", name_prefix, field.name),
+              type{list_type{field.type}},
+            },
+            arrow::ListArray::FromArrays(*combined_offsets, *array).ValueOrDie(),
+          },
+        };
+      },
+      [&](const list_type&) -> indexed_transformation::result_type {
+        if (opt.ignore_lists) {
+          return {{{field.name, field.type}, array}};
+        }
+        return flatten_list(opt, name_prefix, list_offsets, std::move(field),
+                            array);
+      },
+      [&](const record_type&) -> indexed_transformation::result_type {
+        return flatten_record(opt, name_prefix, list_offsets, std::move(field),
+                              array);
+      },
+    };
+    auto result = caf::visit(f, field.type);
+    return result;
+  };
+}
+
+auto make_rename_transformation(std::string new_name)
+  -> indexed_transformation::function_type {
+  return [new_name = std::move(new_name)](struct record_type::field field,
+                                          std::shared_ptr<arrow::Array> array)
+           -> std::vector<std::pair<struct record_type::field,
+                                    std::shared_ptr<arrow::Array>>> {
+    return {
+      {
+        {
+          new_name,
+          field.type,
+        },
+        array,
+      },
+    };
+  };
+}
+
+} // namespace
+
+auto flatten(table_slice slice, flatten_options opt) -> flatten_result {
+  if (slice.rows() == 0)
+    return {std::move(slice), {}};
+  // This version of flatten sometimes breaks if there's a record in a list. We
+  // detect this and warn if in that case.
+  for (const auto& leaf : caf::get<record_type>(slice.schema()).leaves()) {
+    if (const auto* lt = caf::get_if<list_type>(&leaf.field.type)) {
+      if (caf::holds_alternative<record_type>(lt->value_type())) {
+        VAST_WARN("flatten may produce wrong results for records inside lists "
+                  "inside records; this bug will be fixed in a future release");
+        break;
+      }
+    }
+  }
```

Comment:
This is no longer the case, I think.

Reply:
Do you remember what we used to determine this?

For example, I tried out:
`{"foo": [{"bar": "baz", "huh": 2}]}`

`flatten` results in 
`{"foo.bar": ["baz"], "foo.huh": [2]}`

Wasn't that the "broken" behavior?

Reply:
This file in the repository was broken before: vast/integration/data/suricata/rrdata-eve.json

Reply:
I checked the file in flattened and unflattened versions. There is no data loss between the two.

I will carefully remove that message then.